service/project.go

```diff
@@ -211,6 +219,147 @@ func (s *ProjectService) DeleteEnvironment(ctx context.Context, projectID, envID
+		return nil, err
+	}
+
+	_, err := s.Get(ctx, projectID, authUserID)
+	if err != nil {
+		return nil, err
+	}
+
+	i, err := s.invitationRepo.ReadAllForProject(ctx, projectID)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := model.ValidateProjectInvitations(i); err != nil {
+		return nil, apierrors.NewProjectInvitationCorruptedError().Wrap(err)
+	}
+
+	return i, nil
+}
+
+func (s *ProjectService) DeleteInvitation(ctx context.Context, projectID, invitationID, authUserID uuid.UUID) error {
+	if err := s.authSvc.AuthorizeAdminRole(ctx, authUserID); err != nil {
+		return err
+	}
+
+	_, err := s.Get(ctx, projectID, authUserID)
+	if err != nil {
+		return err
+	}
+
+	i, err := s.invitationRepo.Read(ctx, invitationID)
+	if err != nil {
+		switch {
+		case dberrors.IsNotFoundError(err):
+			return apierrors.NewProjectInvitationNotFoundError().Wrap(err)
+		default:
+			return err
+		}
+	}
+
+	return s.invitationRepo.Delete(ctx, i.ID)
+}
+
+func (s *ProjectService) AcceptInvitation(ctx context.Context, tkn cryptox.Token) error {
+	invitation, err := s.getPendingInvitationByToken(ctx, tkn)
+	if err != nil {
+		return err
+	}
+
+	/*
+		When an invitation is confirmed, db function handle_confirmed_invitation is triggered.
+		The function checks if the email is already associated with a user account.
+		If user exists, the function creates a project member and deletes the invitation.
+
+		When user signs up, function handle_new_user is triggered.
+		The function checks if the user's email is associated with a confirmed invitation(s).
+		If so, the function creates a project member(s) and deletes the invitation(s).
+
+		This approach is used because current repository implementation does not support transactions.
+	*/
+	invitation.Accept()
+	return s.invitationRepo.Update(ctx, invitation)
+}
+
+func (s *ProjectService) RejectInvitation(ctx context.Context, tkn cryptox.Token) error {
+	invitation, err := s.getPendingInvitationByToken(ctx, tkn)
+	if err != nil {
+		return err
+	}
+
+	return s.invitationRepo.Delete(ctx, invitation.ID)
+}
+
+func (s *ProjectService) getPendingInvitationByToken(ctx context.Context, tkn cryptox.Token) (model.ProjectInvitation, error) {
+	i, err := s.invitationRepo.ReadByTokenHashAndStatus(ctx, tkn.ToHash(), model.InvitationStatusPending)
+	if err != nil {
+		switch {
+		case dberrors.IsNotFoundError(err):
+			return model.ProjectInvitation{}, apierrors.NewProjectInvitationNotFoundError().Wrap(err)
+		default:
+			return model.ProjectInvitation{}, err
+		}
+	}
+
+	if err := i.Validate(); err != nil {
+		return model.ProjectInvitation{}, apierrors.NewProjectInvitationCorruptedError().Wrap(err)
+	}
+
+	return i, nil
+}
+
+func (s *ProjectService) invitationExists(ctx context.Context, email string, projectID uuid.UUID) (bool, error) {
+	if _, err := s.invitationRepo.ReadByEmailForProject(ctx, email, projectID); err != nil {
+		if dberrors.IsNotFoundError(err) {
+			return false, nil
```

Comment:
Here, I am only checking if the invitation exists; therefore, no validation is necessary.